### PR TITLE
Generic Service State Machine update_status change

### PR DESCRIPTION
**Pull request by tinaafitz**
Modified the update_status method to use new on_error method.

https://www.pivotaltracker.com/story/show/142796573

Related PR https://github.com/ManageIQ/manageiq/pull/14583

### CODE REVIEW

**Comment by tinaafitz** [comment]
@miq-bot add-label services, fine/yes

**Comment by gmcculloug** [comment]
@bzwei Please review

File: spec/content/automate/ManageIQ/Service/Generic/StateMachines/GenericLifecycle.class/__methods__/update_status_spec.rb [hunk 1]
```diff
@@ -65,6 +75,27 @@
 
       msg = "Server [#{miq_server.name}] Service [#{svc_service.name}] Provision Step [] Status [fred] "
       expect(svc_model_request.reload.message).to eq(msg)
+      expect(svc_service).not_to receive(:on_error)
+    end
+  end
+
+  context "on_error provisioning" do
+    let(:root_object) do
```
Comment (bzwei) [inline spec/content/automate/ManageIQ/Service/Generic/StateMachines/GenericLifecycle.class/__methods__/update_status_spec.rb]: You can move `let(:root_object)` out of the context block, and assign `'service_action' => service_action`. Then within each context block you only need to have `let(:service_action)  { 'Provision/Retirement'}`

File: content/automate/ManageIQ/Service/Generic/StateMachines/GenericLifecycle.class/__methods__/update_status.rb [hunk 2]
```diff
@@ -46,24 +46,34 @@ def service_action
...
-              def update_status(service)
-                # Get status from input field status
-                status = @handle.inputs['status']
+              def error
+                @handle.root['ae_result'] == "error"
+              end
 
+              def updated_message(status)
                 updated_message = "Server [#{@handle.root['miq_server'].name}] "
                 updated_message += "Service [#{service.name}] #{service_action} "
```
Comment (mkanoor) [inline content/automate/ManageIQ/Service/Generic/StateMachines/GenericLifecycle.class/__methods__/update_status.rb]: @tinaafitz the service method is being called here to fetch the service object, if someone wants to log another property from the service object, they would call service.some_property that would end up calling the method again. So can we pass the service into this method

File: content/automate/ManageIQ/Service/Generic/StateMachines/GenericLifecycle.class/__methods__/update_status.rb [hunk 3]
```diff
@@ -46,24 +46,34 @@ def service_action
...
+                                            :subject => service,
+                                            :message => "Generic Service Error: #{message}")
+                service.on_error(service_action)
+              end
+
+              def update_status
+                # Get status from input field status
+                status = @handle.inputs['status']
+
+                update_task(updated_message(service, status), status)
```
Comment (bzwei) [inline content/automate/ManageIQ/Service/Generic/StateMachines/GenericLifecycle.class/__methods__/update_status.rb]: Both lines have `update_message(service, status)`. Is it right?

Comment (mkanoor) [inline content/automate/ManageIQ/Service/Generic/StateMachines/GenericLifecycle.class/__methods__/update_status.rb]: @tinaafitz 
We would be calling the service function multiple times here, in case of error there are 3 method calls.
Maybe we can do this in a separate PR. We can either use a @service attribute in the method and just use it everywhere, so we dont have to pass it around or call the function multiple times

**Comment by mkanoor** [comment]
👍

**Comment by simaishi** [comment]
Fine backport details:

```
$ git log -1
commit bfe87c5d55f1c98077b69a3c0f5ea8f859b6095c
Author: Madhu Kanoor <mkanoor@redhat.com>
Date:   Mon Apr 3 10:04:08 2017 -0400

    Merge pull request #85 from tinaafitz/generic_service_on_error
    
    Generic Service State Machine update_status change
    (cherry picked from commit c9739f262ab436bc0142eeab978f929a681b7400)
```
